File: autozig-math/src/isometry2d.rs
use autozig::include_zig;
use crate::{Vec2, Rot2};

#[repr(C)]
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Isometry2d {
    pub rotation: Rot2,
    pub translation: Vec2,
}

include_zig!("zig/isometry2d.zig", {
    fn isometry2d_identity() -> Isometry2d;
    fn isometry2d_new(translation: Vec2, rotation: Rot2) -> Isometry2d;
});

impl Isometry2d {
    pub fn identity() -> Self {
        isometry2d_identity()
    }

    pub fn new(translation: Vec2, rotation: Rot2) -> Self {
        isometry2d_new(translation, rotation)
    }
}

File: autozig-math/zig/isometry2d.zig
const std = @import("std");
const Vec2 = @import("vec2.zig").Vec2;
const Rot2 = @import("rot2.zig").Rot2;

pub const Isometry2d = extern struct {
    rotation: Rot2,
    translation: Vec2,

    pub const IDENTITY = Isometry2d{
        .rotation = Rot2.IDENTITY,
        .translation = .{ .x = 0.0, .y = 0.0 },
    };

    pub fn new(translation: Vec2, rotation: Rot2) Isometry2d {
        return .{
            .translation = translation,
            .rotation = rotation,
        };
    }

    pub fn from_translation(translation: Vec2) Isometry2d {
        return .{
            .translation = translation,
            .rotation = Rot2.IDENTITY,
        };
    }

    pub fn from_rotation(rotation: Rot2) Isometry2d {
        return .{
            .translation = .{ .x = 0.0, .y = 0.0 },
            .rotation = rotation,
        };
    }
};

export fn isometry2d_identity() Isometry2d {
    return Isometry2d.IDENTITY;
}

export fn isometry2d_new(translation: Vec2, rotation: Rot2) Isometry2d {
    return Isometry2d.new(translation, rotation);
}
